docs/book/v3/async-tasks.md

```diff
@@ -0,0 +1,888 @@
+In your local configuration for the server, you'll need to add `task_worker_num`.
+The number of workers you configure define the number of concurrent tasks that can be executed at once.
+Tasks are queued in the order that they trigger, meaning that a `task_worker_num` of 1 will offer no concurrency and tasks will execute in the order they are queued.
+
+```php
+'mezzio-swoole' => [
+    'enable_coroutine' => true, //optional to enable coroutines and useful for tasks coroutines
+    'swoole-http-server' => [
+        'host' => '127.0.0.1',
+        'port' => 8080,
+        'options' => [
+            'worker_num'      => 4,          // The number of HTTP Server Workers
+            'task_worker_num' => 4,          // The number of Task Workers
+            'task_enable_coroutine' => true, // optional to turn on task coroutine support
+        ],
+    ],
+];
+```
+
+> ### No CLI option for task_worker_num
+>
+> Unlike `worker_num`, there is no CLI option for `task_worker_num`.
+> This is because enabling the task worker also requires registering a task worker with the server.
+> To prevent accidental startup failures due to passing an option to specify the number of task workers without having registered a task worker, we omitted the CLI option.
+
+## Task Events
+
+`Mezzio\Swoole\SwooleRequestHandlerRunner` registers listeners on the `Swoole\Http\Server` "task" and "finish" events.  "task" is triggered when `$server->task()` is called, and "finish" is triggered when a task worker calls `$trigger->finish()`.
+Each of the listeners that `SwooleRequestHandlerRunner` registers in turn dispatch an [event via its composed PSR-14 event dispatcher](events.md):
+
+- `Mezzio\Swoole\Event\TaskEvent` is dispatched via the "task" listener.
+- `Mezzio\Swoole\Event\TaskFinishEvent` is dispatched via the "finish" listener.
+
+## Registering Task Listeners
+
+Registering listeners is the same as for other events: you will specify one or more of the above event types, pointing to a list of listeners that are defined as services in your container:
+
+```php
+// in config/autoload/mezzio.global.php:
+
+use Mezzio\Swoole\Event;
+
+return [
+    // ...
+    'mezzio-swoole' => [
+        // ...
+        'swoole-http-server' => [
+            // ...
+            'listeners' => [
+                Event\TaskEvent::class => [
+                    Your\TaskLoggerListener::class,
+                    Your\TaskEventListener::class,
+                ],
+            ],
+        ],
+    ],
+];
+```
+
+> ### TaskFinishEvent listener not required
+>
+> The "finish" event primarily exists to allow you to know when a given task has completed processing.
+> In most cases, you can have a single listener that logs completion of the given task ID, ignoring the return value.
+
+## Shipped task listeners
+
+This package ships two listeners that you can use to process tasks: `Mezzio\Swoole\Task\TaskEventDispatchListener` and `Mezzio\Task\TaskInvokerListener`.
```

Comment:
Please explain a bit about the purpose and differences between TaskEventDispatchListener and TaskInvokerListener. Can or should I bring my own Listener to execute Tasks? Can I use them both or do I have to choose?

I am wondering why a simple `server->task(MyObjectRepresentingATask::createWith($oayload))` and a simple worker listener capable of pulling the task service (MyObjectRepresentingATask::class) from the service manager and invoking it with the given payload isn't provided. You must think that is a bad idea?

Reply:
There's actually three ways to handle tasks:

- Register a listener on the `TaskEvent`, and handle the `TaskEvent` directly. This requires introspecting the result of `$event->getData()` to determine if you can handle it in the first place. As such, it tends to be less type-safe, and thus more brittle.

- Use the `TaskEventDispatchListener`. In this case, the `SwooleRequestHandlerRunner::onTask()` method will invoke that listener, which pulls the data from the payload and treats it as an event to pass to the PSR-14 event dispatcher. You then register listeners agaisnt the various events that will be used as tasks. This is more type safe.

The above two **require** that you queue tasks using `$server->task()`, which means composing the Swoole HTTP Server instance in every class that needs to queue a task. I personally think that's a bad idea.

Why? Because I don't feel your handlers or middleware or services should be aware of the SAPI being used.

So, that leads us to the last approach, the `TaskInvokerListener`. With this one, when you queue a task, you create a `Task` instance that composes both the payload, and the code that will handle it. The `TaskInvokerListener` then checks if `$event->getData()` is a `TaskInterface` instance, and, if so, invokes it.

You _can_ create the task instance manually and pass it to `$server->task()`. But this patch provides an alternate solution: decorating your listener in either `DeferredListener` or `DeferredServiceListener` instances, via the corresponding delegator factories. What does this achieve?

It means you can compose a PSR-14 event dispatcher in your classes instead, and just dispatch events. If a listener can be executed asynchronously, then, in your configuration, you add one of the above delegators to it, and the delegator takes care of creating the appropriate task instance and passing it to the server.

A `DeferredListener` works fine for listeners that have no constructor dependencies or use no objects or resources as dependencies. The `DeferredServiceListener` is good for all the others. Tasks receive the container as an argument when invoked, and the `ServiceBasedTask` will then use that to pull the actual listener from the container and use it.

Now, as for your question of whether you are limited to one listener for tasks or can use multiple: you can use as many as you want. The `TaskEventDispatchListener` assumes that if the payload is an object, it can handle it, and will stop propagation. Similarly, the `TaskInvokerListener` assumes that if it receives a `TaskInterface` in the payload, it can handle it, and will stop propagation.

Personally, though, I'd stick to the `TaskInvokerListener`, and even recommend that later in this particular document.